docs/references/architecture/adr-116-modular-crypto.md

```diff
@@ -0,0 +1,68 @@
+# ADR 116: Modular crypto
+
+## Changelog
+
+ - June 12 2024: Created by @melekes
+
+## Status
+
+**Proposed**
+
+## Context
+
+Currently there's no way to add a new cryptographic curve (used for consensus
+signing) to CometBFT without forking the codebase. Maintaining a fork is a
+significant burden and makes it hard to keep up with the latest changes in the
+main codebase.
+
+On the other hand, **adding a new curve to the main codebase is a significant
+burden as well** because it brings in new dependencies and requires to maintain
+the code for the new curve. For example, the [recent BLS12-381 curve
+addition](https://github.com/cometbft/cometbft/pull/2765).
+
+The goal of this ADR is to make it possible to add new cryptographic curves to
+the CometBFT codebase without forking it, but at the same time avoid the burden
+of maintaining the code for all possible curves in the main codebase.
+
+## Proposal
+
+Add a signature schemes registry to the CometBFT codebase:
+
+```go
+type Signer interface {
+    GenPrivKey() PrivKey
+}
+
+type SignatureScheme uint
+
+const (
+	Ed25519 SignatureScheme = 1 + iota
+	Secp256k1
+	Sr25519
+	Bls12381 // import github.com/cometbft/bls12381
```

Comment:
I'd rather use the _KeyType_ (a string) that every crypto package must define, and which is supposed to be unique. Examples [here](https://github.com/cometbft/cometbft/blob/10bfd89d/crypto/secp256k1/secp256k1.go#L24), [here](https://github.com/cometbft/cometbft/blob/10bfd89d/crypto/ed25519/ed25519.go#L62), or [here](https://github.com/cometbft/cometbft/blob/d27732ad/crypto/bls12381/const.go#L11).